Implement the scriptlevel attribute

eerii opened the pull request:
#### 7f7a5d9197356c9e0116be1bd4e87176d3537fa2
<pre>
Implement the scriptlevel attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=49309">https://bugs.webkit.org/show_bug.cgi?id=49309</a>

Reviewed by Frédéric Wang.

Implements the scriptlevel (<a href="https://w3c.github.io/mathml-core/#dfn-scriptlevel)">https://w3c.github.io/mathml-core/#dfn-scriptlevel)</a>
attribute as defined in the MathML Core specification. Sets a presentational
hint with the corresponding math-depth value.

* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::hasPresentationalHintsForAttribute const):
(WebCore::MathMLElement::collectPresentationalHintsForAttribute):
* Source/WebCore/mathml/mathattrs.in:

Canonical link: <a href="https://commits.webkit.org/302841@main">https://commits.webkit.org/302841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e104a08f81b9c2eb977e2ea382d24792ea661b95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136765 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1575 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1521 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98541 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66437 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132335 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1360 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115889 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79192 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1146 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34021 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80042 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109599 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34521 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139239 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1435 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1384 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107064 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1477 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112234 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106908 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27430 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1166 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30754 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54093 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1506 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64869 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1323 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1360 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,27 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
         else if (equalLettersIgnoringASCIICase(value, "true"_s))
             addPropertyToPresentationalHintStyle(style, CSSPropertyMathStyle, CSSValueNormal);
         return;
+    case AttributeNames::scriptlevelAttr:
```
nit:
```
case AttributeNames::scriptLevelAttr: {
...
}
```

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
         else if (equalLettersIgnoringASCIICase(value, "true"_s))
             addPropertyToPresentationalHintStyle(style, CSSPropertyMathStyle, CSSValueNormal);
         return;
+    case AttributeNames::scriptlevelAttr: {
```
Can you please link to https://w3c.github.io/mathml-core/#dfn-scriptlevel ?

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
         return;
+    case AttributeNames::scriptlevelAttr: {
+        StringView view = value;
+        bool add = value.startsWith('+') || value.startsWith('-');
```
I thinks this is ok, but a bit difficult to read. Maybe this can be a `scriptLevelSyntax` variable instead taking enum values `AddUnsigned`, `AddMinusUnsigned`, `Unsigned`?

eerii replied:
Thank you for the suggestion, it does make it more readable.

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
+    case AttributeNames::scriptlevelAttr: {
+        StringView view = value;
+        bool add = value.startsWith('+') || value.startsWith('-');
+        if (add)
```
```cpp
if (scriptLevelSyntax == AddUnsigned || scriptLevelSyntax == AddMinusUnsigned)
```

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
+        bool add = value.startsWith('+') || value.startsWith('-');
+        if (add)
+            view = view.substring(1);
+        if (!parseInteger<uint64_t>(view, 10, WTF::ParseIntegerWhitespacePolicy::Disallow).has_value())
```
maybe add a comment explaining this is just to valid that the syntax of the unsigned integer.

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
+            view = view.substring(1);
+        if (!parseInteger<uint64_t>(view, 10, WTF::ParseIntegerWhitespacePolicy::Disallow).has_value())
+            return;
+        if (add) {
```
Now this would be a switch

```cpp
switch (scriptLevelSyntax) {
case AddUnsigned:
case AddMinusUnsigned:
   ... use the string builder ...
case
  ...  just do the addPropertyToPresentationalHintStyle ...
```

fred-wang commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -189,6 +191,26 @@ void MathMLElement::collectPresentationalHintsForAttribute(const QualifiedName&
+        if (add) {
+            StringBuilder builder = { };
+            builder.append("add("_s);
+            if (value.startsWith('-'))
```
if (scriptLevelSyntax == AddMinusUnsigned)

fred-wang reviewed (commented):
LGTM. Please consider my suggestion of using an enum if you agree it's more readable.

eerii committed:
Implement the scriptlevel attribute
https://bugs.webkit.org/show_bug.cgi?id=49309

Reviewed by Frédéric Wang.

Implements the scriptlevel (https://w3c.github.io/mathml-core/#dfn-scriptlevel)
attribute as defined in the MathML Core specification. Sets a presentational
hint with the corresponding math-depth value.

* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/css-styling/attribute-mapping-002-expected.txt:
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::hasPresentationalHintsForAttribute const):
(WebCore::MathMLElement::collectPresentationalHintsForAttribute):
* Source/WebCore/mathml/mathattrs.in:

Canonical link: https://commits.webkit.org/302841@main
